### ISSUE TITLE
JS Parsing 2 & Block

### ISSUE THREAD

**Issue by raphpell** [ScintillaOrg/lexilla#53]
```
<script>
	function toJSON(obj){
		if( obj && obj.storePositions ) obj.storePositions()
		var s = JSON.stringify( obj, ['id','from','to','x','y','label','group','OnEnter','OnStay','OnExit','OnTransition'], 2 )
		s = str_replace(
			s,
			[ /,?\s*"([^"]+)":\s*""/gm, /,\s*"/gm, /{\s*/gm, /\s*"([^"]+)":\s*(.+?)\s*([\,}])/gm, /\s*}/gm, /"(\d+)"/gm ],
			[ '', ', "', '{ ', ' $1:$2$3', ' }', '$1' ]
			)
		return s
		}
</script>
```
... no comment !

**Comment by nyamatongwe** [comment]
I can't see a problem. Please explain what you believe is wrong clearly.

**Comment by rdipardo** [comment]
This is a folding bug:

![scite_515_issue53_unfolded](https://user-images.githubusercontent.com/59004801/149606604-df7cf573-f78c-49c8-902a-4d5fb89d5d0d.png)

Looks like a naïve match on the quoted braces is to blame:

![scite_515_issue53_folded](https://user-images.githubusercontent.com/59004801/149606609-2602061a-234f-4c19-82de-2367d8aeadf9.png)



https://github.com/ScintillaOrg/lexilla/blob/66a030c1980ed40e8778ba6752a80bd930d316df/lexers/LexHTML.cxx#L1231-L1232

**Comment by rdipardo** [comment]
On closer inspection, the braces inside the regex literals are changing the fold level, too.

In fact, as long as the folder sees a _balanced_ number of brace pairs, it works as expected:  

```
 2 400   0 + <script>
 2 401   0 + {
 0 402   0 |     /\d{2}/.test('{0}')
 0 402   0 | }
 0 401   0 | </script>
 0 400   0
```

If you count the braces in @raphpell sample, you'll notice they're not balanced:

```
 2 400   0 + <script>
 2 401   0 + {
 0 402   0 |     /\d{2}\}/.test('{0}')
 0 401   0 | }
 0 400   0   </script>
 0 3ff   0
```

**Comment by rdipardo** [comment]
The "umatched" brace inside the regex was the real problem, string states being already filtered by `isStringState()` [^1]. 
Hard to tell at first as the affected code path is quite the dog's breakfast.

Here's a small patch that solely addresses the "reported" issue (assuming my guess is right), with a minimal reproduction for testing: [Issue53.zip](https://github.com/ScintillaOrg/lexilla/files/7878711/Issue53.zip)

To make the folding rules more obvious to the next person, this also reformats a sprawling conditional expression using the [distributive property of Boolean negation](https://en.wikipedia.org/wiki/De_Morgan%27s_laws).

[^1]:  https://github.com/ScintillaOrg/lexilla/blob/66a030c1980ed40e8778ba6752a80bd930d316df/lexers/LexHTML.cxx#L188-L193

**Comment by nyamatongwe** [comment]
Committed fix. Boolean expression is more readable.

**Comment by nyamatongwe** [comment]
Included in 5.1.5 release.
